pkg/discovery/agent.go

```diff
@@ -88,14 +86,14 @@ func (a *Agent) Run() {
 	agent.RegisterAPIValidator(a.validateAPI)
 	agent.OnConfigChange(a.onConfigChange)
 
-	go a.discoveryLoop()
-	go a.publishLoop()
+	go a.discovery.DiscoveryLoop()
+	go a.publisher.PublishLoop()
 
 	select {
 	case <-a.stopAgent:
 		log.Info("Received request to kill agent")
-		a.stopDiscovery <- true
-		a.stopPublish <- true
+		a.discovery.Stop()
```

Comment:
I guess we need to include a scenario for graceful exit as well

Reply:
What does that look like? I thought if the agent is being stopped it is already a graceful exit